libvast/vast/concept/parseable/string/quoted_string.hpp

```diff
@@ -22,22 +22,51 @@
 
 namespace vast {
 
+/// Parser for a string surrounded by a pair of quote characters `Quote`. Inside
+/// the  string, the escape sequence `Esc Quote` can be used to represent a
+/// literal quote, and the escape sequence `Esc Esc` can be used to represent
+/// two occurences of the escape character. All other occurences of the escape
+/// character are interpreted as character literals.
+///
+/// For example, using backslash as escape character and double quote as quote
+/// character:
+///
+///    "foo\n"     denotes the string ['f', 'o', 'o', '\', 'n']
+///    "C:\sys32"  denotes the string ['C', ':', '\', 's', 'y', 's', '3', '2']
+///    "\\"        denotes the string ['\', '\']
+///    "\""        denotes the string ['"']
+///    "\\""       is an invalid string (junk after end of string)
+///    "\\\"       is an invalid string (unterminated)
+
 template <char Quote, char Esc>
 class quoted_string_parser : public parser<quoted_string_parser<Quote, Esc>> {
 public:
   using attribute = std::string;
 
-  static constexpr auto esc = ignore(parsers::ch<Esc>);
-  static constexpr auto quote = ignore(parsers::ch<Quote>);
-  static constexpr auto esc_esc = esc >> parsers::ch<Esc>;
-  static constexpr auto esc_quote = esc >> parsers::ch<Quote>;
-  static constexpr auto str_chr
-    = esc_esc | esc_quote | (parsers::print - quote);
-  static constexpr auto quoted_str = quote >> *str_chr >> quote;
+  static constexpr auto quote = parsers::ch<Quote>;
+  static constexpr auto ignore_esc = ignore(parsers::ch<Esc>);
+  static constexpr auto ignore_quote = ignore(parsers::ch<Quote>);
+  static constexpr auto esc_quote = ignore_esc >> quote;
 
   template <class Iterator, class Attribute>
   bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
-    return quoted_str(f, l, x);
+    // clang-format off
+    if constexpr (std::is_same_v<unused_type, Attribute>) {
+      auto esc_esc = ignore_esc >> ignore_esc;
```

Comment:
You can move the compile time branch into the definition of `esc_esc`:
```c++
    auto esc_esc = [&]() {
      // clang-format off
      if constexpr (std::is_same_v<unused_type, Attribute>)
        return ignore_esc >> ignore_esc;
      else
        // The kleene_parser will append exactly one char per invocation to the
        // output string, so if we returned nothing from this lambda it would
        // append a default-initialized null byte.
        return (ignore_esc >> ignore_esc) ->* [&] { x.push_back(Esc); return Esc; };
      // clang-format on
    }();
```
And avoid the duplication of the following 3 lines.

Reply:
Neat workaround for the new scope introduced by `if constexpr`!